plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -144,20 +154,118 @@ public void processResponseAsync(
                 responseListener.onResponse(response);
                 return;
             }
-            rewriteResponseDocuments(response, responseListener);
+
+            // if many to one, run rewriteResponseDocuments
+            if (!oneToOne) {
+                rewriteResponseDocuments(response, responseListener);
+            } else {
+                // if one to one, make one hit search response and run rewriteResponseDocuments
+                GroupedActionListener<SearchResponse> combineResponseListener = getCombineResponseGroupedActionListener(
+                    response,
+                    responseListener,
+                    hits
+                );
+                AtomicBoolean isOneHitListenerFailed = new AtomicBoolean(false);
+                ;
+                for (SearchHit hit : hits) {
+                    SearchHit[] newHits = new SearchHit[1];
+                    newHits[0] = hit;
+                    SearchResponse oneHitResponse = SearchResponseUtil.replaceHits(newHits, response);
+                    ActionListener<SearchResponse> oneHitListener = getOneHitListener(combineResponseListener, isOneHitListenerFailed);
+                    rewriteResponseDocuments(oneHitResponse, oneHitListener);
+                    // if any OneHitListener failure, try stop the rest of the predictions
+                    if (isOneHitListenerFailed.get()) {
+                        break;
```

Comment:
If breaks here, is the GroupedActionListener can still get enough calls(onResponse/onFailure) to complete?

Reply:
If it breaks, it must go pass line 218, and the combine response listener will onFailure, and that also lead to the response listener onFailure. I also have one to one prediction failed and verify it will only run one prediction only and call onFailure.